File: src/46.全排列.js
/*
 * @lc app=leetcode.cn id=46 lang=javascript
 *
 * [46] 全排列
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  const res = []
  const visited = new Array(nums.length).fill(false)

  function dfs(temp) {
    if (temp.length === nums.length) {
      res.push([...temp])
      return
    }
    for (let i = 0; i < nums.length; i++) {
      if (visited[i]) continue
      visited[i] = true
      temp.push(nums[i])
      dfs(temp)
      temp.pop()
      visited[i] = false
    }
  }
  dfs([])
  return res
};
// @lc code=end

File: src/46.全排列.rs
/*
 * @lc app=leetcode.cn id=46 lang=rust
 *
 * [46] 全排列
 */

// @lc code=start
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut res = vec![];
        let mut nums = nums;


        fn dfs(res: &mut Vec<Vec<i32>>, nums: &mut Vec<i32>, len: usize) {
            if len == nums.len() {
                res.push(nums.to_vec());
                return ;
            }

            for i in len..nums.len() {
                nums.swap(i, len);
                dfs(res, nums, len + 1);
                nums.swap(i, len);
            }
        }
        dfs(&mut res, &mut nums, 0);
        res
    }
}
// @lc code=end

